### PR TITLE
sqlite: exposes SQLite OS Interface "VFS"

### PR DESCRIPTION
**Pull request by d-p-y**
docs: https://www.sqlite.org/vfs.html
It opens ability to do interesting things in v. google: "sqlite vfs based"

### CODE REVIEW

File: vlib/sqlite/sqlite.v [hunk 1]
```diff
@@ -15,10 +15,26 @@ $if windows {
...
+	sqlite_ioerr_delete       = 2570
+
+	sqlite_open_main_db       = 0x00000100
+	sqlite_open_temp_db       = 0x00000200
+	sqlite_open_transient_db  = 0x00000400
+	sqlite_open_main_journal  = 0x00000800
+	sqlite_open_temp_journal  = 0x00001000
+	sqlite_open_subjournal    = 0x00002000
+	sqlite_open_super_journal = 0x00004000
+	sqlite_open_wal           = 0x00080000
```
[the remark this answers is not in the document]
Comment (d-p-y) [reply vlib/sqlite/sqlite.v:37]: constants are taken from official API https://www.sqlite.org/c3ref/constlist.html

[the remark this answers is not in the document]
Comment (spytheman) [reply vlib/sqlite/sqlite.v:37]: I see, so it is correct. There are just some missing constants in between:
(from https://www.sqlite.org/c3ref/c_open_autoproxy.html)
```c
#define SQLITE_OPEN_NOMUTEX          0x00008000  /* Ok for sqlite3_open_v2() */
#define SQLITE_OPEN_FULLMUTEX        0x00010000  /* Ok for sqlite3_open_v2() */
#define SQLITE_OPEN_SHAREDCACHE      0x00020000  /* Ok for sqlite3_open_v2() */
#define SQLITE_OPEN_PRIVATECACHE     0x00040000  /* Ok for sqlite3_open_v2() */
#define SQLITE_OPEN_WAL              0x00080000  /* VFS only */
```

**Comment by d-p-y** [comment]
interesting. When I adapt `sqlite_vfs_lowlevel_test.v` to be regular v program having `test_verify_vfs_is_actually_used` as `main`

```
diff --git a/vlib/sqlite/sqlite_vfs_lowlevel_test.v b/vlib/sqlite/sqlite_vfs_lowlevel_main.v
similarity index 99%
rename from vlib/sqlite/sqlite_vfs_lowlevel_test.v
rename to vlib/sqlite/sqlite_vfs_lowlevel_main.v
index df677eec4..4063cad88 100644
--- a/vlib/sqlite/sqlite_vfs_lowlevel_test.v
+++ b/vlib/sqlite/sqlite_vfs_lowlevel_main.v
@@ -4,7 +4,7 @@ import rand
 const (
        max_file_name_len = 256
 )
-
+/*
 fn test_vfs_register() {
        org_default_vfs := sqlite.get_default_vfs()?

@@ -34,9 +34,9 @@ fn test_vfs_register() {
                panic('vfs supposedly unregistered yet somehow still foundable')
        }
 }
-
+*/
 // minimal vfs based on example https://www.sqlite.org/src/doc/trunk/src/test_demovfs.c
-fn test_verify_vfs_is_actually_used() {
+fn main() {
        wrapped := sqlite.get_default_vfs()?

        vfs_name := 'sometest'
```

and run it `v -cc clang run sqlite_vfs_lowlevel_main.v`
then it works 
```
$v -cc clang run sqlite_vfs_lowlevel_main.v 
fullpathname called
open called
file devicecharacteristics
file devicecharacteristics
file sectorsize
file devicecharacteristics
file sectorsize
file read
file filecontrol
file filecontrol
file devicecharacteristics
file unlock called
file close called
```

**Comment by d-p-y** [comment]
it seems related to garbage collection. When I disable it, test passes under clang

```
$v -cc clang -gc none sqlite_vfs_lowlevel_test.v 
fullpathname called
open called
file devicecharacteristics
file devicecharacteristics
file sectorsize
file devicecharacteristics
file sectorsize
file read
file filecontrol
file filecontrol
file devicecharacteristics
file unlock called
file close called
```

**Comment by spytheman** [comment]
Good find.

**Comment by spytheman** [comment]
try storing `mut vfs_descr := &sqlite.Sqlite3_vfs{` in a global, and add:
```v
[has_globals] 
module main
```
 at the top of the test file

**Comment by d-p-y** [comment]
I did what you proposed

```
diff --git a/vlib/sqlite/sqlite_vfs_lowlevel_test.v b/vlib/sqlite/sqlite_vfs_lowlevel_test.v
index df677eec4..60e4e9c0c 100644
--- a/vlib/sqlite/sqlite_vfs_lowlevel_test.v
+++ b/vlib/sqlite/sqlite_vfs_lowlevel_test.v
@@ -1,3 +1,6 @@
+[has_globals]
+module main
+
 import sqlite
 import rand

@@ -5,6 +8,10 @@ const (
        max_file_name_len = 256
 )

+__global (
+       vfs_descr &ExampleVfsState
+)
+
 fn test_vfs_register() {
        org_default_vfs := sqlite.get_default_vfs()?

@@ -43,7 +50,7 @@ fn test_verify_vfs_is_actually_used() {
        mut vfs_state := &ExampleVfsState{
                log: []string{cap: 100}
        }
-       mut vfs_descr := &sqlite.Sqlite3_vfs{
+       vfs_descr := &sqlite.Sqlite3_vfs{
                iVersion: 2
                szOsFile: int(sizeof(ExampleVfsOpenedFile))
                mxPathname: max_file_name_len
```
and had to do also following thing to deal with mut requirement:

```
diff --git a/vlib/sqlite/vfs_lowlevel.v b/vlib/sqlite/vfs_lowlevel.v
index aec9e59e2..de7a56f0d 100644
--- a/vlib/sqlite/vfs_lowlevel.v
+++ b/vlib/sqlite/vfs_lowlevel.v
@@ -108,7 +108,7 @@ pub fn get_default_vfs() ?&Sqlite3_vfs {
        }
 }

-pub fn (mut v Sqlite3_vfs) register_as_nondefault() ? {
+pub fn (v &Sqlite3_vfs) register_as_nondefault() ? {
```

now both gcc and clang fail when running with gc:
`$v -cc gcc test .`
`$v -cc clang test .`

tcc succeeds with gc `$v -cc tcc test .`

When I run without gc, all pass fine
`$v -cc clang -gc none test .`
`$v -cc gcc -gc none test .`

**Comment by spytheman** [comment]
I think it is definitely a GC bug, related to our prebuilt `thirdparty/tcc/lib/libgc.a`, since it passes with `-d dynamic_boehm` too (i.e. with the system provided `libgc`).

**Comment by spytheman** [comment]
The tricky thing is that bug does not trigger with the clang sanitizers, I think it is related to when the GC decides to do a full collection :-| ...

**Comment by spytheman** [comment]
```
Run ./v missdoc --diff $MOPTIONS pv/vlib vlib
vlib/sqlite/vfs_lowlevel.v:111:0:pub fn (mut v Sqlite3_vfs) register_as_nondefault() ? 
vlib/sqlite/vfs_lowlevel.v:117:0:pub fn (mut v Sqlite3_vfs) unregister() ? 
vlib/sqlite/vfs_lowlevel.v:141:0:pub fn connect_full(path string, mode_flags []OpenModeFlag, vfs_name string) !DB 
vlib/sqlite/vfs_lowlevel.v:100:0:pub fn get_default_vfs() ?&Sqlite3_vfs 
vlib/sqlite/vfs_lowlevel.v:88:0:pub fn get_vfs(name string) ?&Sqlite3_vfs 
Error: Process completed with exit code 1.
```
These new `pub` functions will need to be documented.

**Comment by d-p-y** [comment]
Thank you for help and valuable insight. Will add missing docs tomorrow.

**Pull request merged**